### ISSUE TITLE
iOS 12 App Touch-Areas are offset after keyboard is hidden

### ISSUE THREAD

**Issue by ImperialDevelopment** [ionic-team/capacitor#814]
**Steps to reproduce:**
- Type into an input, which will scroll the Application's view
- Press a button, while the input is focued

**What happens:**
-> The UI shifts down, but touchareas of UI-Elements are still offset. Causing always touching the wrong button

Tested with iPhone X and iPhone 8 Simulator on iOS 11 and iOS 12.
Only happens on iOS 12 (both Simulators)

When you take a look at the View Hierarchy it seems, some Subviews of the WebView are still offset.
Screenshot of Hierarchy attached.

![bildschirmfoto 2018-09-21 um 18 29 11](https://user-images.githubusercontent.com/30909348/45894380-734ae180-bdce-11e8-8d02-ee2c4a8f2484.png)

**Comment by jcesarmobile** [comment]
Are you using Xcode 10, right? 
I think this is a bug in the SDK 12 included in Xcode 10.

Can you try with Xcode 9 and see if you can reproduce there?

**Comment by ImperialDevelopment** [comment]
You are right, does not happen with Xcode 9 / iOS 11.
But when you open an URL and load a website in a WKWebView everything works fine with iOS 12.
So it seems there is an issue in the Capacitor-Layer.

**Comment by jcesarmobile** [comment]
I meant, can you test this in Xcode 9 but on iOS 12 device?

**Comment by ImperialDevelopment** [comment]
Just tried it, built with Xcode 9 and run on iOS 12 device works fine.

**Comment by luke-rogers** [comment]
Issue

I am also seeing this issue when using keyboard which causes input to scroll into view (may be through using keyboard arrows) then closing the keyboard, the buttons in my toolbar cannot be clicked. When you press them nothing happens. This issue is consistently reproducible and makes the app unusable.

Inline with this issue, it does not seem to affect my app when running on iOS 11.

Not super sure the cause of this issue, many issues have been raised across other projects, but it is also effecting capacitor apps.
ionic-team/cordova-plugin-ionic-webview#176
apache/cordova-ios#417

**Comment by nikmartin** [comment]
We just filed the same bug against ionic 4, but was told it was probably a capacitor issue as far as ionic is concerned, spo rather than open another issue, I'll paste it here:



# Bug Report

**Ionic Info**
Run `ionic info` from a terminal/cmd prompt and paste the output below.

```
ionic (Ionic CLI)          : 4.1.0 (/usr/local/lib/node_modules/ionic)
   Ionic Framework            : @ionic/angular 4.0.0-beta.7
   @angular-devkit/core       : 0.7.5
   @angular-devkit/schematics : 0.7.5
   @angular/cli               : 6.1.5
   @ionic/ng-toolkit          : 1.0.8
   @ionic/schematics-angular  : 1.0.6

System:

   NodeJS : v8.11.4 (/usr/local/bin/node)
   npm    : 5.6.0
   OS     : macOS High Sierra
```

**Describe the Bug**
When the on screen keyboard is displayed and hidden (from a footer input), it causes the UI to be scrolled up, the user interface is 'shifted' up by the width of the keyboard. When the keyboard then disappears, all the UI input is still shifted up by the width of the keyboard. This means all clicks in the UI are too high (ie inputs, etc). Also, when an input is in the content area, the keyboard will hide the input instead of scrolling it as it should.

**Steps to Reproduce**
Steps to reproduce the behavior:
1. Clone: https://github.com/grabitlogistics/ios12kbd
2. run npm i && ng build && npx cap add ios && npx cap copy ios && npx cap open ios
3. Run the project on an ios 12 device
4. Click in an upper input to display the keyboard, then click outside to hide the keyboard and lose focus.
5. Click on an input in the footer, making the keyboard appear and shift the inputs up. Click outside the input to hide the keyboard.
6. All touch events have now been shifted up by the height of the keyboard, so to click in the input in the content OR the footer, you must click a few inches above the respective input (about the height of the keyboard)

**Related Code**
sample test case: https://github.com/grabitlogistics/ios12kbd

**Additional Context**
This ONLY happens on iOS 12, on a hardware device OR emulator. To test in an emulator, you must use the emulated keyboard in order to cause the UI to shift up.

**Comment by hghammoud** [comment]
I did some testing on multiple ios/emulators
XCode 10.0

iPhone 8 Plus iOS 11 -> no view shits everything is ok
iPhone 8 Plus iOS 12 -> no view shits everything is ok
IPhone X iOS 11.3 -> view shits but clicks are in place ()
IPhone X ans XS iOS 12 -> view shits and clicks are not accurate (the click action triggers lower the the touch point - same amount of view shit)

it seems related seems related to the extra track pad.
![image](https://user-images.githubusercontent.com/8034427/46759715-5a986200-ccd0-11e8-98cf-84f004e5b16d.png)

**Comment by YodaVN** [comment]
Anyone has the fix of this problem?

**Comment by sbannigan** [comment]
Can confirm that this issue exists in Cordova on iOS 12 as well.

**Comment by KhanhPhamDinh** [comment]
Hi all, i am facing that issue run on IOS 12 that was built with Xcode 10 (if you build with XCode 9, the issue don't appear). IOS 12 will automatic change content offset of WKScrollview in WKWebview fit with height of keyboard in case keyboardWillShow, but it didn't revert it in case keyboardWillHide. That is my solution to fixed it, hope it help for you.

```
/**
 * observer notification when keyboard will hide
 */
[[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(keyboardWillHide)
                                                     name:UIKeyboardWillHideNotification
                                                   object:nil];

/////////////--------------------------//////////////
/*
 *Description: this method was trigger by selector keyboarwillhide from notification
 */
-(void)keyboardWillHide
{
    if (@available(iOS 12.0, *)) {
        WKWebView *webview = (WKWebView*)self.webView;
         for(UIView* v in webview.subviews){
              if([v isKindOfClass:NSClassFromString(@"WKScrollView")]){
                      UIScrollView *scrollView = (UIScrollView*)v;
                      [scrollView setContentOffset:CGPointMake(0, 0)];
              }
          }
     }
}
```

**Comment by JacoRoos** [comment]
@jcesarmobile Regarding this comment: https://github.com/ionic-team/capacitor/issues/814#issuecomment-423609753

Is it not possible to set the SDK that XCode uses to build instead of having to install another version of XCode?

**Comment by jcesarmobile** [comment]
Here it says you can but it present a few problems 
https://stackoverflow.com/questions/44985307/xcode-9-how-to-install-ios-10-sdk

**Comment by sarapoll** [comment]
any solution?

**Comment by hghammoud** [comment]
This was merged in corodva ionic webview 
https://github.com/ionic-team/cordova-plugin-ionic-webview/commit/a67056816ee7fd2b8c74eef1854f2effbd145aea

Anyone knows how to integrate it with capacitor (I am not using cordova plugin)?

**Comment by oliverandersencox** [comment]
not sure if this is related however, my app will also always load first time with a slight gap under the app displaying the WKScrollView:

<img width="498" alt="screenshot 2018-11-21 at 18 20 01" src="https://user-images.githubusercontent.com/6772168/48862579-b02b5880-edbe-11e8-9320-dd9ad14b4265.png">
<img width="456" alt="screenshot 2018-11-21 at 18 34 38" src="https://user-images.githubusercontent.com/6772168/48862580-b02b5880-edbe-11e8-875d-079b4bb2bf7f.png">

If I rerun the app it will sometime disappear, other times it wont - very sporadic

**Comment by snapeuh** [comment]
Hello, we are experiencing the same bug on iOS 12. When we used the keyboard and that it is hidden again the webview keeps a "padding" from the bottom like if the keyboard was still visible.

**Comment by moberwasserlechner** [comment]
Same here. But only after I upgraded to beta.11. 

On every view I use the keyboard the whole app is shifts up and stays there until I rotate my phone to landscape and back. 

Tabbed email input field with keyboard displayed.
![img_3354](https://user-images.githubusercontent.com/1049938/49008023-23d9b680-f16d-11e8-9b0f-585fe34aa12f.PNG)
Keyboard loses focus and disappears but view stays shifted up.
![img_3355](https://user-images.githubusercontent.com/1049938/49008028-250ae380-f16d-11e8-90d2-adf8664ee2f4.PNG)

This is a blocker for all apps using the keyboard.

I experience the problem with XCode 10 and XCode 10.1 deployment target was 11.0.

**Comment by fouzko** [comment]
That last solution works fine !
But I got a problem when I released. The AppDelegate.swift file is not taking into consideration, because of an optimization level during the release build.

That problem can be fixed by disablind the release optimizations :
In Xcode and "Build Settings", seek "optimization level" and select "None [-O0]".

**Comment by rj33** [comment]
As well as the keyboard input issues, I appear to be having essentially the same problem with the UI that pops up to handle selection of choices in a form select element.

I also have a problem with dismissing keyboard and form selector entry elements by clicking in the parent UI, which is probably related.

**Comment by Nudelsieb** [comment]
I had the same issue. My fix was to add the following code to the index.html:

```javascript
    window.addEventListener('keyboardWillHide', () => {
        window.scrollTo(0,0);
	document.body.scrollTop = 0;
    });
```

**Comment by nicdichiara** [comment]
> I had the same issue. My fix was to add the following code to the index.html:
> 
> ```js
>     window.addEventListener('keyboardWillHide', () => {
>         window.scrollTo(0,0);
> 	document.body.scrollTop = 0;
>     });
> ```

Can confirm this works, even if you update xcode to 10.1. Thank you @Nudelsieb !!

**Comment by vuthanhict** [comment]
```

 if #available(iOS 11.0, *) {
             self.webView.scrollView.contentInsetAdjustmentBehavior = .never
 } else {
            self.automaticallyAdjustsScrollViewInsets = false
 }

```

And UIScrollViewDelegate

```
 func scrollViewDidChangeAdjustedContentInset(_ scrollView: UIScrollView) {
       Logger.print("scrollViewDidChangeAdjustedContentInset")
         if #available(iOS 12.0, *) {
           for view in webView.subviews {
                if let scrollView = view as? UIScrollView {
                   scrollView.setContentOffset(.zero, animated: true)
                 }
             }
        }
}
```

Worked

**Comment by christophersansone** [comment]
FWIW, I implemented @Nudelsieb's solution above, with a couple caveats.

1. It requires [cordova-plugin-keyboard](https://github.com/cjpearson/cordova-plugin-keyboard) for the `keyboardWillHide` event to trigger.

2. When immediately transitioning from one input to another, the `keyboardWillHide` event will trigger, even though the keyboard does not actually have time to hide.  That seemed to cause undesired scrolling.  I added a `setTimeout()` to prevent it from executing immediately, which helped.

```javascript
const scrollToTopFn = () => {
  if (window.Keyboard && !window.Keyboard.isVisible) {
    window.scrollTo(0,0);
    window.document.body.scrollTop = 0;
  }
};

window.addEventListener('keyboardDidHide', () => {
  window.setTimeout(scrollToTopFn, 100);
});
```

**Comment by thestevenmellor** [comment]
Where did you place that code? Which file? I'm running into the same touch area bug after using input

**Comment by zenochan** [comment]
1. template
```html
    <input  (blur)="scrollToTop()">
```

2. page
```ts
  scrollToTop()
  {
    window.scrollTo(0,0);
  }
```

3. down

**Comment by sooheesh** [comment]
https://github.com/ionic-team/cordova-plugin-ionic-webview/issues/176#issuecomment-455032360

this worked.

**Comment by longbo666** [comment]
期待有人能真正解决

**Comment by daniellizik** [comment]
> You are right, does not happen with Xcode 9 / iOS 11.
> But when you open an URL and load a website in a WKWebView everything works fine with iOS 12.
> So it seems there is an issue in the Capacitor-Layer.

I can still reproduce this with iOS 11.4.1 and xcode 9.4.1

**Comment by jcesarmobile** [comment]
I think this should be fixed in next release, I made a few keyboard changes that combined seem to fix it.

@daniellizik what you comment should be a different issue as this only affects iOS 12 when using Xcode 10, not iOS 11 despite the Xcode version, and doesn't affect any iOS version when using Xcode 9.

**Comment by alphagamer7** [comment]
@jcesarmobile Confirming that even with the latest release (4.0.2) the issue is not fixed.

**Comment by jcesarmobile** [comment]
Not sure what 4.0.2 you mean, but I’m talking about next capacitor release (1.0.0 beta 18, still not released)

**Comment by Stylemayster** [comment]
@jcesarmobile Is the fix that you've made is in a branch that we can test?

**Comment by jcesarmobile** [comment]
Beta 18 is out, can you test with it?

**Comment by solojuve1897** [comment]
@jcesarmobile
Part of this bug is fixed but there is still strange behaviour. Check my video:

https://drive.google.com/file/d/11Msg2lSiW1UoN-si6nL1QxLS93shBu9h/view

None of these occur on my Android-devices.

**Comment by calvinckho** [comment]
@jcesarmobile it is working great on my iPhone 7 test device running iOS 12! Thanks for this update.

**Comment by davidrinnan** [comment]
when the given page loads, if it is ios etc etc, I attach a listener 
document.addEventListener('focusout', resetScroll);

reset scroll is running on a short timeout and checks if the innerheight and height is the same, if se, it resets the scroll. This results in the scroll only being reset once the keyboard is removed from screen, rather than everytime the focus is changed. 
```
 setTimeout(()=>{
        if (window.innerHeight == screen.height) {
            window.scrollTo(0, 0)    
        }
    },timeout);
```

**Comment by jcesarmobile** [comment]
I can't reproduce since beta 18 and calvinckho also confirmed that it's working fine for him, so I'm going to close.
For other strange behaviours report new issues.
